### PR TITLE
[AOTInductor] Add a C tensor struct and C inferface

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105331

Summary: Introduce a C tensor struct and tensor operation APIs to avoid
ABI compatibility issues in the AOTInductor generated binary.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

Differential Revision: [D47680387](https://our.internmc.facebook.com/intern/diff/D47680387)

### CODE REVIEW

**Comment by vadimkantorov** [comment]
btw I wonder, can existing DLPack structures/dtypes be sufficient structures for the C interface/

**Comment by desertfire** [comment]
Use this PR to collect feedbacks. I will commit another one from fbcode later.

File: torch/csrc/inductor/aot_inductor_tensor.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    void* aten_tensor);
+
+EXPORT AotInductorTensor aot_inductor_empty_strided(
+    int64_t dim,
+    int64_t* sizes_ptr,
+    int64_t* strides_ptr,
+    AotInductorDevice device,
+    AotInductorScalarType type);
+
+EXPORT AotInductorTensor aot_inductor_as_strided(
```
Comment (jansel) [inline torch/csrc/inductor/aot_inductor_tensor.h:62]: You may want to cherry-pick the `torch.ops.inductor._reinterpret_tensor` part of #102625 and use that instead of burning the name `as_strided` into this API.

There is actually a subtle bug in how TorchInductor uses `as_strided`.  The difference between the two is:
- as_strided discards the storage_offeset() of the input tensor
- _reinterpret_tensor adds the new offset to the storage_offeset() of the input tensor

So they act the same on a tensor with a `storage_offset=0`, but differ when `storage_offset>0`.

Basically whenever inductor generates as_strided, it really means _reinterpret_tensor.  It should be possible to construct a test case that triggers this bug, by chaining offsets with more complex views.

Currently, this doesn't bite us in practice, because usually `storage_offset=0`.  However, when we start doing memory planning it will become an issue.

File: torch/csrc/inductor/aot_inductor_tensor.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    AotInductorTensor self,
+    int64_t* sizes_ptr,
+    int64_t* strides_ptr,
+    int64_t offset) {
+  c10::ArrayRef sizes(sizes_ptr, self.ndim);
+  c10::ArrayRef strides(strides_ptr, self.ndim);
+  at::Tensor self_tensor;
+  convert_to_aten_tensor(self, &self_tensor);
+  at::Tensor result = at::as_strided(self_tensor, sizes, strides, offset);
+  return convert_to_aot_inductor_tensor(&result);
```
Comment (jansel) [inline torch/csrc/inductor/aot_inductor_tensor.cpp:165]: We can implement this directly on AotInductorTensor and get much better perf than calling at::as_strided.

File: torch/csrc/inductor/aot_inductor_tensor.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      c10::TensorOptions()
+          .device(convert_to_c10_device(inductor_tensor.device))
+          .dtype(convert_to_c10_scalar_type(inductor_tensor.type));
+
+  at::Tensor* t = static_cast<at::Tensor*>(aten_tensor);
+  (*t) = at::from_blob(
+      inductor_tensor.data_ptr,
+      c10::IntArrayRef(inductor_tensor.sizes, inductor_tensor.ndim),
+      c10::IntArrayRef(inductor_tensor.strides, inductor_tensor.ndim),
+      options);
```
Comment (jansel) [inline torch/csrc/inductor/aot_inductor_tensor.cpp:136]: What is the memory management story here?  Who owns the memory?  How do we ensure it doesn't get freed before this tensor falls out of scope?

My understanding is `from_blob` doesn't have a refcount on the storage.

File: torch/csrc/inductor/aot_inductor_tensor.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    AotInductorDevice device,
+    AotInductorScalarType type) {
+  c10::ArrayRef sizes(sizes_ptr, dim);
+  c10::ArrayRef strides(strides_ptr, dim);
+  at::Tensor result = at::empty_strided(
+      sizes,
+      strides,
+      c10::TensorOptions(convert_to_c10_device(device))
+          .dtype(convert_to_c10_scalar_type(type)));
+  return convert_to_aot_inductor_tensor(&result);
```
Comment (jansel) [inline torch/csrc/inductor/aot_inductor_tensor.cpp:152]: I believe this results in a pointer to freed memory.

File: torch/csrc/inductor/aot_inductor_tensor.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    AotInductorDeviceType_to_C10DeviceType = {
+        {kAotInductorCPU, c10::kCPU},
+        {kAotInductorCUDA, c10::kCUDA},
+};
+
+std::unordered_map<c10::DeviceType, AotInductorDeviceType>
+    C10DeviceType_to_AotInductorDeviceType = {
+        {c10::kCPU, kAotInductorCPU},
+        {c10::kCUDA, kAotInductorCUDA},
+};
```
Comment (jansel) [inline torch/csrc/inductor/aot_inductor_tensor.cpp:59]: switch statements might be faster than `std::unordered_map`

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/inductor/aot_inductor_tensor.cpp (hunk 5 above)
Comment (jansel) [reply torch/csrc/inductor/aot_inductor_tensor.cpp:59]: An even better option might be to use the exact same IDs as eager pytorch, then have some `static_assert`s in this file to make sure they stay in sync.  That way we can convert for free.

File: torch/csrc/inductor/aot_inductor_tensor.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/inductor/aot_inductor_tensor.cpp:136]: The semantics here is that at::Tensor is a non-owning references to the AotInductorTensor, which sounds pretty difficult to use correctly. You probably need to pass around a function pointer that knows how to do deallocation and setup some move semantics, or arrange that the model so never returns any allocations (the runtime is expected to do all the allocations, and aot inductor only borrows.)

Comment (desertfire) [reply torch/csrc/inductor/aot_inductor_tensor.cpp:136]: One solution is to combine this with memory planning (https://github.com/pytorch/pytorch/pull/102625), so basically memory planner is our runtime wrt. memory.

**Comment by desertfire** [comment]
@jansel @ezyang , I have another crazy idea. Instead of introducing an AotInductorTensor struct, how about just calling all fallback ops with wrapper functions that passe input/output tensors using `void*` pointers? The questions about memory management are still relevant, but seems like a C struct is not adding extra value here?

**Comment by vadimkantorov** [comment]
About design, maybe onnxruntime's scheme of managing memory / pre allocation and in general - can be useful / worth checking

File: torch/csrc/inductor/aot_inductor_tensor.cpp (hunk 3 above)
Comment (jansel) [reply torch/csrc/inductor/aot_inductor_tensor.cpp:136]: That would almost work, but there are fallback kernels like at::convolution() that don't have `out=` variants, so they return a newly-allocated tensors.

I think you need something like a `std::shared_ptr<AotTensorStorage>` to keep the memory allocated and reference count when it can be deleted.  Either that or you could manually increment/decrement the reference count on the at::TensorStorage.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47680387).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47680387).

**Comment by desertfire** [comment]
There may be some CI issues to be fixed, but it's ready for another round of reviews, @ezyang  @jansel

File: torch/csrc/inductor/aot_inductor_utils.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <assert.h>
+#include <stdint.h>
+#include <memory>
+#include <stdexcept>
+#include <string>
```
Comment (ezyang) [inline torch/csrc/inductor/aot_inductor_utils.h:7]: This file is very confusing to me. Is this part of the C ABI? Why are there C++ headers???

File: torch/csrc/inductor/aot_inductor_tensor.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  kAotInductorBFloat16 = 15,
+} AotInductorScalarType;
+
+typedef struct {
+  AotInductorDeviceType device_type;
+  int8_t device_id;
+} AotInductorDevice;
+
+// Bare minumum tensor struct to interchange data between AOTInducotor
+// generated shared library and libtorch to avoid ABI compatibility issues
```
Comment (ezyang) [inline torch/csrc/inductor/aot_inductor_tensor.h:39]: If you want other people to be able to work on this file, you're going to need a more detailed description about what exactly the ABI guarantees are, and under what circumstances you can change the declarations in this file. This is especially important because (1) we don't have that many C++ experts on the project and (2) we have even fewer people who regularly have to deal with ABI stable code. fbcode peeps are even worse, since they are spoiled by the monorepo setup. The default assumption should be that someone touching this code has no idea about the ABI concerns and you need big flashing warnings on this file (and any OTHER files) about what is and not OK to do.

File: torch/csrc/inductor/aot_inductor_tensor.h [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+typedef struct {
+  void* data_ptr;
+  void* deleter;
+  AotInductorDevice device;
+  AotInductorScalarType dtype;
+  int64_t ndim;
+  const int64_t* sizes;
+  const int64_t* strides;
+  int64_t storage_offset;
+} AotInductorTensor;
```
Comment (ezyang) [inline torch/csrc/inductor/aot_inductor_tensor.h:49]: If these structs are part of the public ABI, I am a bit nervous about them. You have no padding on the structs, so it's not possible to add new fields to these structs without it being an ABI breaking change. A common alternative is to provide an opaque function for allocating and filling the struct, which you can then successively upgrade the functions (`_v2`, etc) if you need to add new fields later.

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
+#pragma once
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:1]: So everything in c/ is the public headers?

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+// surface that could cause ABI breakage. AOTInductor generated code will
+// NOT be allowed to call any of aten/c10 C++ APIs. It is only allowed to call
+// functions declared in this header file instead.
+
+// WARNING: Change the following types can break ABI compatibility
+// Typedef these to be exactly the same as c10 since C enum types can not be
+// specified as int8_t. We will assert the data type length at run time.
+typedef int8_t AOTInductorDeviceType;
+typedef int8_t AOTInductorDeviceIndex;
+typedef int8_t AOTInductorScalarType;
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:22]: To be honest, it's still not clear at all to me why we need to pass device type/index / scalar type around. In particular, we're going to have a thrift sidecar for the FFI calls right? So we can also use that to do argument type checking entirely on the runtime side.

Comment (ezyang) [reply torch/csrc/aot_inductor/c/aten_shim.h:22]: Also, IMO, there isn't really an reason to use a narrower type here. You're only passing these in as arguments, but that doesn't change the ABI at all https://godbolt.org/z/TG44PrWjj

And the downside to def'ing them too narrow is that if you ever do need to expand it, it's very painful.

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+#define kAOTInductorQInt8 12
+#define kAOTInductorQUInt8 13
+#define kAOTInductorQInt32 14
+#define kAOTInductorBFloat16 15
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// WARNING: Change the following signatures can break ABI compatibility
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:51]: not can, will ;)

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+AOT_INDUCTOR_EXPORT void aot_inductor_initialize();
+
+AOT_INDUCTOR_EXPORT void aot_inductor_destroy();
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:54]: It's be helpful to know what the lifecycle requirements for initialize/destroy are in the comments here

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+AOT_INDUCTOR_EXPORT void aot_inductor_free(AOTInductorTensorHandle aot_tensor);
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:56]: It's more conventional to define an unspecified struct and define a pointer to it, e.g., `AOTInductorTensor*`. This makes it clear that the type in question is a pointer and, e.g., needs manual memory management. When you elide the pointer, it is easy to accidentally think (esp in a context where C++ is available) that this is a proper struct with RAII.

Comment (ezyang) [reply torch/csrc/aot_inductor/c/aten_shim.h:56]: Also, we need comments on all functions saying what the ownership semantics of the pointer arguments are

File: torch/csrc/aot_inductor/c/utils.h [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+#else // !__GNUC__
+#ifdef _WIN32
+#define AOT_INDUCTOR_EXPORT __declspec(dllexport)
+#else // !_WIN32
+#define AOT_INDUCTOR_EXPORT
+#endif // _WIN32
+#endif // __GNUC__
+
+#define AOT_INDUCTOR_CHECK(cond, msg) \
+  { assert((cond) && (msg)); }
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/utils.h:17]: Still not sure why this needs to be in the public headers. Also still not sure about the use of C assert construct

Comment (ezyang) [reply torch/csrc/aot_inductor/c/utils.h:17]: In particular, assert will be compiled unless you do a debug build (which we don't by default)

File: torch/csrc/aot_inductor/c/aten_shim.h (hunk 13 above)
Comment (ezyang) [reply torch/csrc/aot_inductor/c/aten_shim.h:56]: I noticed that the XXXHandle convention is what AITemplate is doing, I guess we can keep it as is.

Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:56]: The naming here is weird. `aot_inductor` is the prefix for the namespace (since C doesn't have namespaces). So this function is effectively just `free`. Free what? Put the noun in!

Comment (ezyang) [reply torch/csrc/aot_inductor/c/aten_shim.h:56]: (This applies to other functions as well below, I won't call them out explicitly)

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+AOT_INDUCTOR_EXPORT void* aot_inductor_data_ptr(
+    AOTInductorTensorHandle aot_tensor);
+
+AOT_INDUCTOR_EXPORT AOTInductorTensorHandle
+convert_input_output_to_aot_tensor(AtenTensorHandle aten_tensor);
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:62]: AtenTensorHandle seems like a strange type to put into a header that must be ABI compatible. What's going on here? Is this actually a function that must be used SOLELY from the runtime (non-model.so) and interacts with the public C ABI to produce an AOTInductorTensorHandle? If so, it shouldn't be in this header.

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 16]
```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <stdint.h>
+#include <torch/csrc/aot_inductor/c/utils.h>
+
+// What problems are we trying to solve here with a C shim layer?
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/aot_inductor/c/aten_shim.h:6]: From a naming perspective, it would be good to distinguish these functions from the model.so ABI functions. In particular, `aot_inductor_` as a naming prefix seems extremely non-descriptive, as it refers to the compiler, whereas we generally want to talk about the predictor service / runtime, as opposed to the model.so

To get the naming ball rolling, I suggest: shortening `aot_inductor` to `aoti`. Next, we split the function namespaces into `aoti_runtime` and `aoti_model`. This file is aoti_runtime only.

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+      AOT_INDUCTOR_CHECK(false, "Unkown AOTInductorScalarType")
+      return 0;
+  }
+}
+
+// TODO: this probably can also be statically planned to avoid any dynamic
+// allocation
+thread_local std::vector<std::unique_ptr<AOTInductorTensor>> tensor_manager_;
+#ifndef NDEBUG
+thread_local std::unordered_set<void*>* live_buffer_;
```
Comment (ezyang) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:138]: So what's this? Is this a way to make sure we aren't leaking tensors?

File: torch/csrc/aot_inductor/c/aten_shim.h (hunk 12 above)
Comment (ezyang) [reply torch/csrc/aot_inductor/c/aten_shim.h:54]: `aoti_runtime_initialize` makes it a lot clearer what is being initialized here

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+      return 16;
+    default:
+      AOT_INDUCTOR_CHECK(false, "Unkown AOTInductorScalarType")
+      return 0;
+  }
+}
+
+// TODO: this probably can also be statically planned to avoid any dynamic
+// allocation
+thread_local std::vector<std::unique_ptr<AOTInductorTensor>> tensor_manager_;
```
Comment (ezyang) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:136]: Instead of shoving this in a thread local, it seems like it would more explicit to represent the runtime state as an explicit struct and explicitly pass it around. This doesn't seem like a big burden to the model.so, which is all code-genned anyway. Otherwise, you need to clearly document the threading requirements (specifically, model.so is never allowed to migrate to another thread; you are not allowed to run multiple runtimes on the same thread, etc)

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 19]
```diff
@@ -0,0 +1,492 @@
...
+          .options(options)
+          .make_tensor();
+}
+
+} // namespace
+
+void aot_inductor_initialize() {
+  // c10 types and enums should rarely change, but use this checking to
+  // fail loudly in the worst case
+  AOT_INDUCTOR_CHECK(
```
Comment (ezyang) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:250]: You are in runtime, you have access to all ATen headers, there is no need to use a completely unrelated error checking macro?

That being said, you have sort of kicked up a beehive here. The problem is that it is not defined to raise exceptions across a C ABI boundary. See also https://stackoverflow.com/questions/6426835/what-happens-if-exception-gets-thrown-through-c-code The frame unwinding mechanism is particularly delicate and it would be much more obviously safe if we avoid throwing exceptions this way. To do this, we must explicitly return error codes when an exception would have been raised from this file, and then the model must propagate this exception in some way. The simplest implementation would lose all exception information, but since you will eventually return the runtime, the full error information could likely be stored in a thread local. If it is a burden on codegen to explicitly check all of the error codes, you could choose to turn an error code into an exception inside the model.so (but once again, once you get back to the runtime-model interface, you need to convert back into a C error again.)

**Comment by ezyang** [comment]
Not really anywhere close to done reviewing. Maybe you want to schedule some pairing sessions.

**Comment by suo** [comment]
random drive by thought, maybe good to bring https://lvc.github.io/abi-compliance-checker into our ci

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    AOTInductorTensorHandle mat1,
+    AOTInductorTensorHandle mat2,
+    float beta,
+    float alpha,
+    uint8_t use_gelu);
+
+AOT_INDUCTOR_EXPORT AOTInductorTensorHandle aot_inductor_bmm_out(
+    AOTInductorTensorHandle out,
+    AOTInductorTensorHandle self,
+    AOTInductorTensorHandle mat2);
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:98]: Concretely, proposed new signature:

```
// invariant DeviceGuard on the CUDA id already
void aoti_runtime_bmm_float16_3d_cuda(
  float16* out, // b x n x p
  float16* self, // b x n x m (must be contiguous)
  float16* mat2, // b x m x p (must be contiguous)
  int64_t b,
  int64_t n,
  int64_t m,
  int64_t p,
);
```

(pick a consistent ordering for the adjectives

(ignoring error code for a moment)

File: torch/csrc/aot_inductor/c/aten_shim.h [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+    AOTInductorDeviceType device_type,
+    AOTInductorDeviceIndex device_index,
+    AOTInductorScalarType dtype);
+
+AOT_INDUCTOR_EXPORT AOTInductorTensorHandle aot_inductor_as_strided(
+    AOTInductorTensorHandle self,
+    int64_t ndim,
+    const int64_t* sizes_ptr,
+    const int64_t* strides_ptr,
+    int64_t storage_offset);
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/aten_shim.h:77]: This shouldn't be necessary!

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 22]
```diff
@@ -0,0 +1,492 @@
...
+      static_cast<AOTInductorScalarType>(t->scalar_type()));
+}
+
+AOTInductorTensorHandle aot_inductor_empty_strided(
+    int64_t ndim,
+    const int64_t* sizes_ptr,
+    const int64_t* strides_ptr,
+    AOTInductorDeviceType device_type,
+    AOTInductorDeviceIndex device_index,
+    AOTInductorScalarType dtype) {
```
Comment (ezyang) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:379]: Assume that device guard is set appropriately (who is setting the device guard? modifying the cuda runtime state, which is shared)

`aoti_runtime_cuda_alloc(int64_t nbytes)`

File: torch/_inductor/codegen/wrapper.py [hunk 23]
```diff
@@ -621,6 +632,14 @@ def codegen_python_shape_tuple(self, shape: Tuple[Expr, ...]) -> str:
     def codegen_shape_tuple(self, shape: Tuple[Expr, ...]) -> str:
         return self.codegen_python_shape_tuple(shape)
 
+    def codegen_as_strided(self, name, size, stride, offset, indent_buffer=None) -> str:
+        size = self.codegen_shape_tuple(size)
+        stride = self.codegen_shape_tuple(stride)
+        offset = self.codegen_sizevar(offset)
+        if offset != "0":
+            return f"as_strided({name}, {size}, {stride}, {offset})"
+        return f"as_strided({name}, {size}, {stride})"
```
Comment (ezyang) [inline torch/_inductor/codegen/wrapper.py:641]: ?

File: torch/csrc/aot_inductor/c/model_interface.h [hunk 24]
```diff
@@ -59,9 +51,9 @@ AOTInductorError AOTInductorModelContainerDelete(
 // Runs the inference.
 AOTInductorError AOTInductorModelContainerRun(
     AOTInductorModelContainerHandle container_handle,
-    const AOTInductorTensorHandle inputs_handle,
+    AtenTensorHandle* inputs_handle,
     size_t num_inputs,
-    AOTInductorTensorHandle outputs_handle,
+    AtenTensorHandle* outputs_handle,
     size_t num_outputs,
     AOTInductorStreamHandle stream_handle);
```
Comment (ezyang) [inline torch/csrc/aot_inductor/c/model_interface.h:58]: Maybe we need to phone in @chenyang78, but I don't understand model_interface.h at all. Naively, I expected this to be the C ABI which you can access model.so from the runtime. However, I would NOT expect it to be a plain list of functions, because I am dlopen'ing the model.so and I am getting a pile of function pointers that I would probably assign into a struct to dynamic dispatch into. So I'd expect a bunch of function pointer signatures, not... this. What the heck is this?

It is pointless for the header file that predictor service links against to be ABI compatible if the implementation is linked into predictor, because you can assume that this will always be compiled together, so you can make sure they're synchronized in lockstep. Am I misunderstanding the situation here?

Comment (chenyang78) [reply torch/csrc/aot_inductor/c/model_interface.h:58]: @ezyang The function pointers are declared in the internal runtime header file, AOTInductorModelImpl.h. It is included in the source file (AOTInductorModelImpl.cpp) where we dlopen the model.so. This was how the AIT was originally designed so that I just re-used a similar mechanism. I think one difference is that AIT also open-sources the corresponding runtime part of the code (https://github.com/facebookincubator/AITemplate/blob/main/fx2ait/fx2ait/csrc/AITModelImpl.h), so it's probably less confused to some extent.

Comment (ezyang) [reply torch/csrc/aot_inductor/c/model_interface.h:58]: Yes, please open source this code lol

Comment (ezyang) [reply torch/csrc/aot_inductor/c/model_interface.h:58]: @chenyang78 Then, in that case, model_interface.h doesn't actually have any C ABI requirement, and it shouldn't be placed in the c/ folder

Comment (chenyang78) [reply torch/csrc/aot_inductor/c/model_interface.h:58]: @ezyang I think we have two sets of C interfaces, one declared in aten_shim.h and one declared in model_interface.h. The former is supposed to be used internally in the aot_inductor_model.so for mitigating ABI compatibility issue.  The latter is used by the external user code that loads aot_inductor_model.so. 

Is the c/ folder meant for holding only internal ABI-related files? If Yes, we should probably move model_interface.h to somewhere else. On the other hand, if the c/ folder is the place for all C interface files, it might be fine to leave model_interface.h there?

File: torch/_inductor/codegen/aot_inductor_model_interface.cpp [hunk 25]
```diff
@@ -1,16 +1,18 @@
...
 #include <stdexcept>
 #include <vector>
+#include <iostream>
 
 #define CONVERT_EXCEPTION_TO_ERROR_CODE(...)     \
   try {                                          \
     __VA_ARGS__                                  \
   } catch (const std::exception& e) {            \
-    LOG(ERROR) << "Error: " << e.what();         \
+    std::cerr << "Error: " << e.what();         \
```
Comment (khabinov) [inline torch/_inductor/codegen/aot_inductor_model_interface.cpp:12]: Nit: new line symbol would be needed here.

File: torch/_inductor/codegen/aot_inductor_model_interface.cpp [hunk 26]
```diff
@@ -1,16 +1,18 @@
...
     return AOTInductorError::Failure;            \
   } catch (...) {                                \
-    LOG(ERROR) << "Unknown exception occurred."; \
+    std::cerr << "Unknown exception occurred."; \
```
Comment (khabinov) [inline torch/_inductor/codegen/aot_inductor_model_interface.cpp:15]: Nit: new line symbol would be needed here.

File: torch/_inductor/codegen/aot_inductor_model_interface.cpp [hunk 27]
```diff
@@ -21,7 +23,7 @@ AOTInductorError AOTInductorModelContainerCreate(
     AOTInductorModelContainerHandle* container_handle,
     size_t num_models) {
   if (num_models == 0) {
-    LOG(ERROR) << "num_models must be positive, but got 0";
+    std::cerr << "num_models must be positive, but got 0";
```
Comment (khabinov) [inline torch/_inductor/codegen/aot_inductor_model_interface.cpp:26]: Nit: new line symbol would be needed here.

File: torch/_inductor/codegen/wrapper.py [hunk 28]
```diff
@@ -836,7 +858,7 @@ def codegen_allocation(self, buffer):
             AllocateLine(
                 self,
                 buffer,
-                not self.use_preallocated_ouput(buffer),
+                not self.use_preallocated_ouput(name),
```
Comment (khabinov) [inline torch/_inductor/codegen/wrapper.py:861]: Nit: `use_preallocated_output`

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 29]
```diff
@@ -0,0 +1,492 @@
...
+#include <ATen/ops/from_blob.h>
+#include <ATen/ops/mm.h>
+
+#endif
+
+// This file will be compiled as a part of libtorch, and the AOTInductor
+// compiled model binary will load these C functions from the deployment host.
+// As explained in aten_shim.h, this helps to solve ABI compatibility issues.
+//
+// In addtion, there is another benefit of adding a shim layer: for AOTInductor
```
Comment (khabinov) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:27]: Nit: In addition

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 30]
```diff
@@ -0,0 +1,492 @@
...
+// generated fallback aten calls, we don't have to go through its regular heavy
+// machinery to maintain aten tensors and call through dispatacher. For
```
Comment (khabinov) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:29]: dispatacher -> dispatcher

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 31]
```diff
@@ -0,0 +1,492 @@
...
+// instance, we could implement a fast path to call device-specific mm library
+// instead of calling at::mm as a fallback. TODO: this fuctionality is largely
+// not done yet.
+
+namespace {
+
+// An internal struct to represent a tensor in AOTInductor
+//
+// This is NOT meant to be a general replacement of at::Tensor. The design
+// choices made here are highly couple with AOTInductor's codegen, especially
```
Comment (khabinov) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:39]: couple -> coupled?

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 32]
```diff
@@ -0,0 +1,492 @@
...
+      data_ptr,
+      deleter,
+      ndim,
+      sizes,
+      strides,
+      storage_offset,
+      device_type,
+      device_index,
+      dtype));
+  return tensor_manager_[tensor_manager_.size() - 1].get();
```
Comment (khabinov) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:161]: Nit: `tensor_manager_.back().get()`

File: torch/csrc/aot_inductor/aten_shim_common.cpp [hunk 33]
```diff
@@ -0,0 +1,492 @@
...
+  void print() {
+    std::cout << "data_ptr=" << data_ptr << ", ";
+    std::cout << "deleter=" << deleter << ", ";
+    std::cout << "ndim=" << ndim << ", ";
+    std::cout << "sizes=" << sizes << ", ";
+    std::cout << "strides=" << strides << ", ";
+    std::cout << "storage_offset=" << storage_offset << ", ";
+    std::cout << "device_type=" << int32_t(device_type) << ", ";
+    std::cout << "device_index=" << int32_t(device_index) << ", ";
+    std::cout << "dtype=" << int32_t(dtype) << std::endl;
```
Comment (khabinov) [inline torch/csrc/aot_inductor/aten_shim_common.cpp:100]: Nit: maybe one big `std::cout`?

File: torch/csrc/aot_inductor/c/utils.h (hunk 14 above)
Comment (khabinov) [reply torch/csrc/aot_inductor/c/utils.h:17]: Yeah, without `NDEBUG` all asserts will be wiped out by the preprocessor

**Comment by desertfire** [comment]
Closed in favor of https://github.com/pytorch/pytorch/pull/109022
